crates/compiler/can/src/traverse.rs

```diff
@@ -981,3 +989,54 @@ pub fn find_declaration(symbol: Symbol, decls: &'_ Declarations) -> Option<Found
         }
     }
 }
+
+pub fn find_declaration_at(region: Region, decls: &Declarations) -> Option<FoundDeclaration<'_>> {
+    let mut visitor = Finder {
+        region,
+        found: None,
+    };
+
+    visitor.visit_decls(decls);
+    return visitor.found;
+
+    struct Finder<'a> {
+        region: Region,
+        found: Option<FoundDeclaration<'a>>,
+    }
+
+    impl Visitor for Finder<'_> {
+        fn should_visit(&mut self, region: Region) -> bool {
+            region.contains(&self.region)
+        }
+
+        fn visit_decl(&mut self, decl: DeclarationInfo<'_>) {
+            if self.should_visit(decl.region()) {
+                match decl {
+                    DeclarationInfo::Value { .. }
+                    | DeclarationInfo::Function { .. }
+                    | DeclarationInfo::Destructure { .. } => {
+                        self.found = Some(FoundDeclaration::Decl(unsafe {
```

Comment:
I've reused the code just above this for a similar purpose but it includes some unsafe. I'm fairly confident that it's all safe but I wanted to highlight it for review.

this is the code I repurposed from just above this in the file
https://github.com/roc-lang/roc/blob/544a5dcfec8bc96cafbbb888bc3e39f0f8a6dcb2/crates/compiler/can/src/traverse.rs#L953-L978